snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+	minStragglerCheckInterval time.Duration
+
+	// log logs events
+	log logging.Logger
+
+	// minConfirmationThreshold is the minimum stake percentage that below it, we do not check if we are stragglers.
+	minConfirmationThreshold float64
+
+	// connectedPercent returns the stake percentage of connected nodes.
+	connectedPercent func() float64
+
+	// connectedValidators returns a set of tuples of NodeID and corresponding weight.
+	connectedValidators func() set.Set[ids.NodeWeight]
+
+	// lastAcceptedByNodeID returns the last accepted height a node has reported, or false if it is unknown.
+	lastAcceptedByNodeID func(id ids.NodeID) (ids.ID, bool)
+
+	// processing returns whether this block ID is known and its descendants have not yet been accepted by consensus.
+	// This means that when the last accepted block is given as input, true is returned, as by definition
+	// its descendants have not been accepted by consensus, but this block is known.
+	// For any block ID belonging to an ancestor of the last accepted block, false is returned,
+	// as the last accepted block has been accepted by consensus.
+	processing func(id ids.ID) bool
+
+	// lastAccepted returns the last accepted block of this node.
+	lastAccepted func() ids.ID
+
+	// getSnapshot returns a snapshot of the network's nodes and their last accepted blocks,
+	// or false if it fails from some reason.
+	getSnapshot func() (snapshot, bool)
+
+	// haveWeFailedCatchingUp returns whether we have not replicated enough blocks of the given snapshot
+	haveWeFailedCatchingUp func(snapshot) bool
+}
+
+type stragglerDetector struct {
+	stragglerDetectorConfig
+
+	// continuousStragglingPeriod defines the time we have been straggling continuously.
+	continuousStragglingPeriod time.Duration
+
+	// previousStragglerCheckTime is the last time we checked whether
+	// our block height is behind the rest of the network
+	previousStragglerCheckTime time.Time
+
+	// prevSnapshot is the snapshot from a past iteration.
+	prevSnapshot snapshot
+}
+
+func newStragglerDetector(
+	getTime func() time.Time,
+	log logging.Logger,
+	minConfirmationThreshold float64,
+	lastAccepted func() (ids.ID, uint64),
+	connectedValidators func() set.Set[ids.NodeWeight],
+	connectedPercent func() float64,
+	processing func(id ids.ID) bool,
+	lastAcceptedByNodeID func(ids.NodeID) (ids.ID, bool),
+) *stragglerDetector {
+	sd := &stragglerDetector{
+		stragglerDetectorConfig: stragglerDetectorConfig{
+			lastAccepted:              dropHeight(lastAccepted),
+			processing:                processing,
+			minStragglerCheckInterval: minStragglerCheckInterval,
+			log:                       log,
+			connectedValidators:       connectedValidators,
+			connectedPercent:          connectedPercent,
+			minConfirmationThreshold:  minConfirmationThreshold,
+			lastAcceptedByNodeID:      lastAcceptedByNodeID,
+			getTime:                   getTime,
+		},
+	}
+
+	sd.getSnapshot = sd.getNetworkSnapshot
```

Comment:
I also think replacing the pointers as part of the initialization is ugly. I can divide the struct into several pieces so that it would look less ugly.

However, just swapping the function pointers with interfaces won't make the code any less bug-prone.